<a href="https://colab.research.google.com/github/JNestman/cse380-notebooks/blob/master/PonderProve07.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Explore Fermat’s Little Theorem Further
Fermat’s Little Theorem (FLT) says that if N is prime, then N divides XN - X.


Remember, the contrapositive of the conditional statement in this theorem is that if N doesn’t divide XN - X for some X, then N can’t be prime.
Unfortunately, this simple primality test doesn’t always work, because it can be fooled by pseudoprimes.

341 = 11 · 31 is not prime. But 341 does divide 2341 - 2.

print((pow(2, 341) - 2) % 341)

0

So 341 is a base-2 pseudoprime. What about base-3?

print((pow(3, 341) - 3) % 341)

165

So 341 is not a base-3 pseudoprime.

Are there any other bases that will not fool the FLT test for 341?

Are there any pseudoprimes that will fool the FLT for any choice of base coprime to them?

Yes! Your task is the find the first 4-digit bullet-proof pseudoprime (bppp) and prove
(yes, PROVE) that it will fool the FLT test for every base coprime to it. Your proof must
use all of the following:
1. the definition of coprime,
2. a consequence of coprimality,
3. the factorization of the bppp,
4. FLT, and the
5. CRT (Chinese Remainder Theorem).


In [1]:
# Actually, there are only 9000 4-digit numbers,
# so how hard can it be to look at all of them?

# But even better, since we want the smallest *bppp*,
# we can stop as soon as we find it!

from math import gcd
from sympy import isprime

def passes_FLT_test_even_though_not_prime(b, n):
  # primes don't count as pseudoprimes
  return not isprime(n) and (b ** n) % n == b

def is_bppp(n):
  bases_coprime_to_n = [b for b in range(2, n) if gcd(b, n) == 1]
  return all(list(map(lambda b: passes_FLT_test_even_though_not_prime(b, n), 
                      bases_coprime_to_n)))

n = 1000
while not is_bppp(n):
  n += 1

n

1105

# Proof



1.   The function produces 1105 as the first four digit **bppp** and we are out to prove that 1105 will fool the FTL test for every base coprime to it.
2.   1105 is a composite number, 1105 = 5 * 13 * 17

3. For some number $n$, if gcd( n , 1105) = 1, then gcd( n , 5) = gcd( n , 13) = gcd( n , 17) = 1.

4. By FML these facts follow:
  1. $n^{4} \equiv_5 1$
  2. $n^{12} \equiv_{13} 1$
  3. $n^{16} \equiv_{17} 1$

5. Because 1104 is 1 less than 1105, and is also a multiple of 4, 12, and 16 --- which are 1 less than 5, 13, and 17 --- it follows that:

$(n^4)^{276} = n^{1104} \equiv_{5} 1$

$(n^{12})^{92} = n^{1104} \equiv_{13} 1$

$(n^{16})^{69} = n^{1104} \equiv_{17} 1$

Therefore, $b^{1104} \equiv_{1105} 1$.

6. 1105 and 1104 are coprime numbers.

7. The Chinese remainder theorem states that if the $n_i$ are coprime, and if $a_1$, ..., $a_k$ are integers such that 0 ≤ $a_i$ < $n_i$ for every $i$, then there is one and only one integer $x$, such that 0 ≤ $x$ < $N$ and the remainder of the Euclidean division of $x$ by $n_i$ is $a_i$ for every $i$

8. It follows that every base coprime of 1105 will also fool the the FTL test.
